crates/matrix-sdk/src/send_queue.rs

```diff
@@ -698,6 +698,32 @@ impl QueueStorage {
         Ok(true)
     }
 
+    /// Replace an event that has been sent with
+    /// [`Self::push`] with the given transaction id, before it's been actually
+    /// sent.
+    ///
+    /// Returns whether the given transaction has been effectively edited. If
+    /// false, this either means that the transaction id was unrelated to
+    /// this queue, or that the event was sent before we edited it.
+    async fn replace(
+        &self,
+        transaction_id: &TransactionId,
+        serializable: SerializableEventContent,
+    ) -> Result<bool, RoomSendQueueStorageError> {
+        if self.being_sent.read().await.contains(transaction_id) {
+            return Ok(false);
+        }
+
+        self.client
+            .get()
+            .ok_or(RoomSendQueueStorageError::ClientShuttingDown)?
+            .store()
+            .update_send_queue_event(&self.room_id, transaction_id, serializable)
+            .await?;
```

Comment:
Can we have a race condition here? I think no, but maybe `being_sent` can be modified before `update_send_queue_event` is called?

Reply:
No indeed: there's only one possible caller, the sender task. Will add a small comment.